### DISCUSSION TITLE
Add access to NextURL in server components

### DISCUSSION THREAD

**Discussion by Yoshie2000** [vercel/next.js#49369] [Ideas]
### Goals

There currently seems to be no way to access the requested URL (including search params etc.) in server components.

Client components have `useRouter()` and `useSearchParams()` hooks, but server components don't have anything like that.

### Background

I wanted to create a sidenav layout for my application, with the current link being active. The plan was to make it a server component, because there would be no reactivity involved.

Instead I had to make the sidenav a client component and use something like `usePathname()` to highlight the active link on the client side. This made the whole component a lot more complex that it needed to be.

### Proposal

My idea is to implement a function like `pathname()` / `url()` that returns a `NextURL`, similar to the existing functions `cookies()` or `headers()` that work in all server components.

Alternatively, the URL could be passed to (all / layout / page) components via props.

I haven't contributed to NextJS yet, but for me this is a very important feature and I'd be happy to contribute towards it's development.

**Comment by Yoshie2000** [comment]
I started a draft pull request for this feature: https://github.com/vercel/next.js/pull/49533

**Comment by sophiebits** [comment]
In the current design it's intentional that you cannot access the full URL from layout components. The reason is in order to make sure that we can reuse the layout when navigating between pages in the common case and only fetch the new inner page content. There are a couple options:

1. Use a client component that reads usePathname(), as you mentioned
2. Add a parallel route like `app/@sideNav/[...path]/page.tsx`. This will refetch when the path changes and you can render it from the root layout component.
3. Split your routes so they have different layout files depending on what needs to be shown in the side nav (for just highlighting the active link this doesn't make sense, but it could in other cases).

Remember that client components are an integral part of the server components architecture. Using them isn't a bad thing. In this case it means that the sidebar content does not need to be refetched when navigating; instead the component can be downloaded once in the JS bundle and then update what it renders based on its current state.

**Reply by Yoshie2000** [reply]
Hey, thanks for your reply!

I can definitely get behind that point, but I have another situation where I'm not sure what the intended approach would be:

Imagine a dashboard application with a filter by year. All the data in the application depends on this filter, meaning every fetch needs to know the selected year to fetch the correct data. Even some links in the nav bar depend on it.

The filter would presumably be described by a search param for easy accessibility.

Is there a better way to do this than to fetch everything client side? I feel like that kind of defeats the purpose of SSR, and would result in loading spinner hell.

**Reply by sophiebits** [reply]
If you have a search param ?year=2023 that is conceptually part of the root layout then you can use a middleware URL rewrite to put it at the beginning of the path by the time your app handles it. Then the root layout will be refetched when the year changes but not when navigating between pages with the same year.
